src/java.base/share/classes/java/lang/String.java

```diff
@@ -3272,14 +3273,29 @@ static String join(String prefix, String suffix, String delimiter, String[] elem
 
         int off = 0;
         prefix.getBytes(value, off, coder); off += prefix.length();
-        if (size > 0) {
-            var el = elements[0];
+        if (size > 1) {
+            String el = elements[0];
             el.getBytes(value, off, coder); off += el.length();
-            for (int i = 1; i < size; i++) {
-                delimiter.getBytes(value, off, coder); off += delimiter.length();
+            final int delimiterBytesBegin = off << coder;
+            delimiter.getBytes(value, off, coder); off += delimiterLength;
+            el = elements[1];
+            el.getBytes(value, off, coder); off += el.length();
+            final int delimiterLengthWithCoder = (delimiterLength << coder);
+            for (int i = 2; i < size; i++) {
+                System.arraycopy(
+                        value,
+                        delimiterBytesBegin,
+                        value,
+                        off << coder,
+                        delimiterLengthWithCoder
```

Comment:
Is there any benefit to special casing short delimiters?  1..4 bytes?
Arraycopy has a fixed overhead regardless of the delimiter length.
I might expect a big payoff for coder = Latin1 and delimiter length == 1..2  compared to arraycopy.